dbms/src/Interpreters/JoinPartition.cpp

```diff
@@ -540,32 +550,81 @@ void NO_INLINE insertBlockIntoMapsTypeCase(
         segment_index_info[segment_index].push_back(i);
     }
 
-    bool null_need_materialize = isNullAwareSemiFamily(current_join_partition->getJoinKind());
-    for (size_t insert_index = 0; insert_index < segment_index_info.size(); insert_index++)
+    std::list<size_t> insert_indexes;
+    for (size_t i = 0; i < segment_index_info.size(); ++i)
+    {
+        size_t insert_index = (i + stream_index) % segment_index_info.size();
+        insert_indexes.emplace_back(insert_index);
+    }
+
+#define INSERT_TO_MAP(join_partition, segment_index)                                                                       \
+    auto & current_map = (join_partition)->getHashMap<Map>();                                                              \
+    for (auto & i : (segment_index))                                                                                       \
+    {                                                                                                                      \
+        Inserter<STRICTNESS, Map, KeyGetter>::insert(current_map, key_getter, stored_block, i, pool, sort_key_containers); \
+    }
+
+#define INSERT_TO_NOT_INSERTED_MAP                                                                      \
+    /* null value */                                                                                    \
+    /* here ignore mutex because rows_not_inserted_to_map is privately owned by each stream thread */   \
+    /* for right/full out join or null-aware semi join, need to insert into rows_not_inserted_to_map */ \
+    assert(rows_not_inserted_to_map != nullptr);                                                        \
+    assert(segment_index_info.size() == (1 + segment_size));                                            \
+    bool null_need_materialize = isNullAwareSemiFamily(current_join_partition->getJoinKind());          \
+    for (auto index : segment_index_info[segment_size])                                                 \
+    {                                                                                                   \
+        rows_not_inserted_to_map->insertRow(stored_block, index, null_need_materialize, pool);          \
+    }
+
+    // First use tryLock to traverse twice to find all segments that can acquire locks immediately and execute insert.
+    //
+    // If there is only one segment left, there is no need to use try locks
```

Comment:
How about set a up limit of try_lock loop times, if the try_lock loop time exceeds the limit, we fallback to blocking lock?